src/rhsmlib/facts/network.py

```diff
@@ -0,0 +1,176 @@
+# Copyright (c) 2023 Red Hat, Inc.
+#
+# This software is licensed to you under the GNU General Public License,
+# version 2 (GPLv2). There is NO WARRANTY for this software, express or
+# implied, including the implied warranties of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+# along with this software; if not, see
+# http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+#
+# Red Hat trademarks are not licensed under GPLv2. No permission is
+# granted to use or replicate Red Hat trademarks that are incorporated
+# in this software or its documentation.
+import json
+import logging
+import socket
+import subprocess
+from typing import Callable, Dict, List, Literal, Union
+
+from rhsmlib.facts import collector
+
+log = logging.getLogger(__name__)
+
+
+class NetworkCollector(collector.FactsCollector):
+    def __init__(
+        self,
+        arch: str = None,
+        prefix: str = None,
+        testing: bool = None,
+        collected_hw_info: Dict[str, Union[str, int, bool, None]] = None,
+    ):
+        super().__init__(arch=arch, prefix=prefix, testing=testing, collected_hw_info=None)
+
+        self.hardware_methods: List[Callable] = [
+            self.get_network,
+            self.get_interfaces,
+        ]
+
+    def _query_ip_command(self) -> List[dict]:
+        """Call system's 'ip' command and return its value as a dictionary."""
+        output = subprocess.run(["ip", "--json", "address"], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        if output.stderr != b"":
+            log.error(f"Could not query 'ip' for network facts: {output.stderr}")
+            return []
+        stdout: str = output.stdout.decode("utf-8")
+        return json.loads(stdout)
+
+    def _extract_address_list(self, family: Literal["inet", "inet6"], data: List[dict]) -> List[str]:
+        """Extract IPv4 or IPv6 addresses from 'ip' output.
+
+        The list excludes loopback and link-local addresses.
+        """
+        result: List[str] = []
+        for interface in data:
+            if "LOOPBACK" in interface["flags"]:
+                continue
+            result += [
+                address["local"]
+                for address in interface["addr_info"]
+                if address["family"] == family and address["scope"] != "link"
+            ]
+        return result
+
+    def get_network(self) -> dict:
+        """Get general network facts.
+
+        Hostname, FQDN and a list of IPv4/IPv6 addresses get collected here.
+
+        Resulting facts have 'network.' prefix.
+        """
+        # These socket functions work even with no network available
+        data = self._query_ip_command()
+
+        result: Dict[str, str] = {
+            "network.hostname": socket.gethostname(),
+            "network.fqdn": socket.getfqdn(),
```

Comment:
there is a comment currently in `HardwareCollector.get_network_info()` about why `socket.getfqdn()` is not used:
https://github.com/candlepin/subscription-manager/blob/2d980dc37ebbeb20ddf30f6af019a5e48e1f575d/src/rhsmlib/facts/hwprobe.py#L760-L783
also I remember something about old bugs wrt this in the history, so i'd check that out; in the worst case, that code can be kept in its own private method